### PR TITLE
feat: adding support for multiqueue

### PR DESCRIPTION
**Pull request by drew-viles**
*Issue #, if available:*
#501 

*Description of changes:*
Right now, users cannot add queues to the network NIC. With this PR, users will now be able to optionally supply a value into the CAPI provider to configure queues.

*Testing performed:*
Added tests for `make test` and ran with success. I'm currently trying to get access to a proxmox with shared storage as it appears to be required for this where as right now, I'm using local storage. 

If you want to wait for that testing to be done on my side I totally get it! However, the change seems pretty straight forward on the whole so I'll let you make the call from your side.

### CODE REVIEW

**Comment by wikkyk** [comment]
Nice and simple change but we'll hold it for now until after we have released 0.8.0 and v1alpha2.

File: api/v1alpha1/proxmoxmachine_types.go [hunk 1]
```diff
@@ -405,6 +405,11 @@ type NetworkDevice struct {
 	// +kubebuilder:validation:Maximum=4094
 	VLAN *uint16 `json:"vlan,omitempty"`
 
+	// Queues is the Multiqueue field in Proxmox to assign to the NIC
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	Queues *uint16 `json:"queues,omitempty"`
```
Comment (wikkyk) [inline api/v1alpha1/proxmoxmachine_types.go]: Fields must be signed. If you're expecting 65535 to be a valid value, then make it an int32 with a Maximum rule, if 65534 then int16 is fine.

File: api/v1alpha1/proxmoxmachine_types.go [hunk 2]
```diff
@@ -405,6 +405,11 @@ type NetworkDevice struct {
 	// +kubebuilder:validation:Maximum=4094
 	VLAN *uint16 `json:"vlan,omitempty"`
 
+	// Queues is the Multiqueue field in Proxmox to assign to the NIC
```
Comment (wikkyk) [inline api/v1alpha1/proxmoxmachine_types.go]: It'd be nice to make this comment a bit more informative:
```suggestion
	// Queues is the number of queues assigned to the device.
	// This value is passed to the Multiqueue field in PROXMOX.
```

File: internal/service/vmservice/utils.go [hunk 3]
```diff
@@ -108,6 +108,21 @@ func extractNetworkVLAN(input string) uint16 {
 	return 0
 }
 
+// extractNetworkQueue returns the queue out of net device input e.g. virtio=A6:23:64:4D:84:CB,bridge=vmbr1,mtu=1500,tag=100,queue=4.
+func extractNetworkQueue(input string) uint16 {
```
Comment (wikkyk) [inline internal/service/vmservice/utils.go]: Remember to fix the return value to match the Queues field.

This increases the importance of #653 :-)

File: internal/service/vmservice/utils.go [hunk 4]
```diff
@@ -179,14 +194,21 @@ func shouldUpdateNetworkDevices(machineScope *scope.MachineScope) bool {
 				return true
 			}
 		}
+
+		if v.Queues != nil {
+			queue := extractNetworkQueue(net)
+			if queue != *v.Queues {
```
Comment (wikkyk) [inline internal/service/vmservice/utils.go]: This could be done with ptr.Deref.

File: internal/service/vmservice/utils.go [hunk 5]
```diff
@@ -179,14 +194,21 @@ func shouldUpdateNetworkDevices(machineScope *scope.MachineScope) bool {
...
 	}
 
 	return false
 }
 
 // formatNetworkDevice formats a network device config
-// example 'virtio,bridge=vmbr0,tag=100'.
-func formatNetworkDevice(model, bridge string, mtu *uint16, vlan *uint16) string {
+// example 'virtio,bridge=vmbr0,tag=100,queue=4'.
+func formatNetworkDevice(model, bridge string, mtu *uint16, vlan *uint16, queue *uint16) string {
```
Comment (wikkyk) [inline internal/service/vmservice/utils.go]:
```suggestion
// example 'virtio,bridge=vmbr0,tag=100,queues=4'.
func formatNetworkDevice(model, bridge string, mtu *uint16, vlan *uint16, queues *uint16) string {
```

**Review by wikkyk — CHANGES_REQUESTED** [review]
Please port this to v1alpha2!
Remember to update the tests with the correct type, too.

**Comment by drew-viles** [comment]
Sorry absolutely slammed atm, I will pick this up again asap though

**Comment by wikkyk** [comment]
v0.8.0 has been released so we are reopening feature merges. Please port your PR to v1alpha2 and address my comments and we'll do another review. This is a good feature, I would really like it to go in.

**Comment by drew-viles** [comment]
I'll looks at getting it done this weekend. Time has been not been on my side the past few weeks.

**Comment by drew-viles** [comment]
ok, I think I've got everything covered, `make test` is working nicely again so let me know if you have any thoughts on this.

File: api/v1alpha2/proxmoxmachine_types_test.go [hunk 6]
```diff
@@ -367,6 +367,20 @@ var _ = Describe("ProxmoxMachine Test", func() {
 		})
 	})
 
+	Context("Queues", func() {
+		It("Should not allow machine with network device queue equal to 0", func() {
```
Comment (wikkyk) [inline api/v1alpha2/proxmoxmachine_types_test.go:371]: Please add a test for `>65535`, a test for `0<queues<65536`, and a test for `nil`.

File: internal/service/vmservice/utils.go [hunk 7]
```diff
@@ -180,14 +195,19 @@ func shouldUpdateNetworkDevices(machineScope *scope.MachineScope) bool {
...
 	}
 
 	return false
 }
 
 // formatNetworkDevice formats a network device config
-// example 'virtio,bridge=vmbr0,tag=100'.
-func formatNetworkDevice(model, bridge string, mtu *int32, vlan *int32) string {
+// example 'virtio,bridge=vmbr0,tag=100,queues=4'.
+func formatNetworkDevice(model, bridge string, mtu *int32, vlan *int32, queues int32) string {
```
Comment (wikkyk) [inline internal/service/vmservice/utils.go]: This function takes pointers to optional values, please follow that pattern.

**Review by wikkyk — CHANGES_REQUESTED** [review]

**Review by wikkyk — CHANGES_REQUESTED** [review]
Also please note the linter failures.

**Comment by drew-viles** [comment]
ok, addressed those points. 🤞 Hopefully that's good to go now.

**Comment by wikkyk** [comment]
Nice, I really like it now. As this is a new feature and I'm intending for v0.8.2 to be primarily a bugfix release, I'll formally approve and schedule this for merging into v0.8.3.

**Comment by drew-viles** [comment]
~ooh that test is suggesting I need to hit `make generate/make mockgen`. I've done it locally but there are no changes. Am I missing something?~

Edit:
Nevermind, `make all` did the trick ;-) 

Thanks!

**Comment by drew-viles** [comment]
I've just noticed all of my commits except the last one are **unverified**. I'm not sure why that happened tbh as I have signing on by default - something obviously got skipped when I did them 🤷 . If you want me to sign them all and re-push them let me know. Happy to do so!

**Comment by wikkyk** [comment]
It's because *I* rebased your commits, so technically they're new commits which you haven't signed.
I'll squash this whole PR into one commit, anyway.

Also the plans for v0.8.2 have changed, merging this now pending test results.

**Review by wikkyk — APPROVED** [review]

**Comment by wikkyk** [comment]
Test coverage is lower than expected because the extractor is uncovered. (Actually, it just doesn't seem reachable but it is tested indirectly; were it not working, proxmoxmachine_types tests would fail.) I'm ok with this because of #653.

**Comment by wikkyk** [comment]
Thanks!
